---
title: "Overview of mice KO data"
author: "Carlos Ruiz"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```


## Introduction

In this document, I will show an overview of the data downloaded from the MGI database.

```{r}
library(tidyverse)
library(ontologyIndex)

load("../results/preprocess/final_mice_genotype_phenotype.Rdata")
mpo_ont <- get_ontology("../data/MGI/MPheno_OBO.ontology")

```

## Genotype data

We preprocessed the data downloaded from MGI. We ensured that *all the alleles were Knock-out*. This criteria were also applied to combinations of genes, so all genes should be modified with knock-out alleles. We also considered wild-type alleles, resulting in heterozygous genotypes.

The same gene could have different alleles in different genetic backgrounds. Each combination of gene alleles and genetic background were considered as an independent sample.

MGI data also reported alleles with no effect on the phenotype. We selected those genes where none of their alleles impacted on the phenotype, and added them genes to the final dataset. These genes were considered as homozygous for the knock-out by default.

```{r}
all_genos <- unlist(strsplit(genotype_phenotype$MGIMarkerGenotype, ","))
all_genes <- unlist(strsplit(all_genos, "/"))
all_genes <- all_genes[grep("MGI", all_genes)]
```


Overall, our final dataset contains `r nrow(genotype_phenotype)` combinations of gene alleles and genetic backgrounds corresponding to `r length(unique(all_genes))` different genes. `r sum(is.na(genotype_phenotype$PhenotypeID))` combinations of genes corresponded to alleles without phenotypic effect.

```{r}
no_effect <- subset(genotype_phenotype, TopPhenotypeIDs == "")$MGIMarkerGenotype 
data.frame(Type = c("Effect on phenotype", "No effect"),
           N = c(length(unique(all_genes)) - length(unique(no_effect)), 
                 length(unique(no_effect)))) %>%
  ggplot(aes(x = Type, y = N)) +
    geom_bar(stat = "identity") +
  xlab("Gene type") +
  theme_bw()
```

Most genes had an effect on the phenotype, while `r length(unique(no_effect))` genes did not produce any effect.

Genes were present in combinations with very different frequencies. 

```{r}
df_gene_freq <- table(table(all_genes)) %>% data.frame() %>%
  mutate(N = as.numeric(Var1),
         Group = cut(N, 
                     c(0, 1, 2, 3, 5, 10, 200), 
                     c("1", "2", "3", "4-5", "6-10", "10+"))) %>%
  group_by(Group) %>%
  summarize(N = sum(Freq))
ggplot(df_gene_freq, aes(x = Group, y = N)) +
  geom_bar(stat = "identity")
```

Most of the genes were only present in one combination of alleles. However, we found some genes present in more than 50 gene combinations.

```{r}
marker_list <- strsplit(subset(genotype_phenotype, !is.na(PhenotypeID))$MGIMarkerGenotype, ",")
n_hom <- sapply(marker_list, function(x){
  sum(unlist(strsplit(x, "/", fixed = TRUE)) == "2")
})
tibble(N_comb = factor(lengths(marker_list)),
       N_hom = factor(n_hom)) %>%
  ggplot(aes(x = N_comb, fill = N_hom)) +
    geom_bar() +
    theme_bw() +
  xlab("Genes in combination") +
  scale_fill_discrete(name = "Homozygous genes")
    

tibble(N_comb = lengths(marker_list),
       N_hom = factor(n_hom)) %>%
  filter(N_comb > 2) %>%
  mutate(N_comb = factor(N_comb)) %>%
  ggplot(aes(x = N_comb, fill = N_hom)) +
    geom_bar() +
    theme_bw() +
  xlab("Genes in the combination") +
  scale_fill_discrete(name = "Homozygous genes")
    

```

Most of the allele combination only had one gene. Nonetheless, we had combinations with up to 7 genes. In general, most combinations were composed of knock-out alleles in homozygosity.

## Phenotype data

Allele combinations were associated to different phenotypes using the Mammalian Phenotype Ontology (MPO). The MPO is a phenotype ontology which relates the phenotypes hierarchicaly. For the sake of simplicity, we mapped all phenotypes to any of the 28 phenotypes in the level 2 of the hierarchy. Importantly, each phenotype can be children of multiple level 2 phenotypes and each allele combination can affect multiple phenotypes. Therefore, each allele combination can be mapped to multiple level 2 phenotypes.

```{r}
all_phenos <- strsplit(genotype_phenotype$TopPhenotypeIDs, ",")

pheno_tab <- table(unlist(all_phenos))
names(pheno_tab) <- sapply(names(pheno_tab), function(x){
  mpo_ont$name[mpo_ont$id == x]
})

tibble(Phenotype = names(pheno_tab), N = as.vector(pheno_tab)) %>%
  arrange(desc(N)) %>%
  knitr::kable()

```

There is a great variability in the frequency of the phenotypes, ranging from sensory phenotypes produced by hundreds of allele combinations to mortality or metabolism phenotypes produced by around 10K allele combinations.

```{r}
pheno_freq <- table(lengths(all_phenos)) 
tibble(Freq = names(pheno_freq), 
       N = as.vector(pheno_freq)) %>%
  mutate(Freq = factor(Freq, levels = names(pheno_freq))) %>%
  ggplot(aes(x = Freq, y = N)) +
  geom_bar(stat = "identity") +
  theme_bw()
```
Most allele combinations affected one phenotype. However, a high proportion of allele combination affected multiple phenotypes. 


```{r}
tibble(N_genes = lengths(marker_list),
       N_phenos = lengths(all_phenos)) %>%
  filter(N_phenos > 0) %>%
  ggplot(aes(x = factor(N_genes), y = N_phenos)) +
  geom_boxplot() +
  theme_bw() +
  xlab("Number of genes") +
  ylab("Number of phenotypes")
  
```

There is no correlation between the number of genes knocked-out and the number of phenotypes affected.


